File: lib/_models/user_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class User {
  final String id;
  final String address;
  final String birthday;
  final String course;
  final int deleted_at;
  final String email;
  final String first_name;
  final String gender;
  final String last_name;
  final String phone;
  final int role_id;
  final int school_id;
  final String username;
  final String year;
  final String photoURL;

  User({
    required this.id,
    required this.address,
    required this.birthday,
    required this.course,
    required this.deleted_at,
    required this.email,
    required this.first_name,
    required this.gender,
    required this.last_name,
    required this.phone,
    required this.role_id,
    required this.school_id,
    required this.username,
    required this.year,
    required this.photoURL,
  });

  Map<String, dynamic> toMap() {
    return {
      'id' : id,
      'address' : address,
      'birthday' : birthday,
      'course' : course,
      'deleted_at' : deleted_at,
      'email' : email,
      'first_name' : first_name,
      'gender' : gender,
      'last_name' : last_name,
      'phone' : phone,
      'role_id' : role_id,
      'school_id' : school_id,
      'username' : username,
      'year' : year,
      'photoURL' : photoURL
    };
  }

  @override
  String toString() {
    return "{'id' : ${id}, 'email' : ${email}, 'first_name' : ${first_name}, 'gender' : ${gender}, 'school_id' : ${school_id}, 'username' : ${username}, 'year' : ${year} }";
  }

  User.fromDocumentSnapshot(DocumentSnapshot<Map<String, dynamic>> userMap)
    : id = userMap.id,
      address = userMap.data()!["address"],
      birthday = userMap.data()!["birthday"],
      course = userMap.data()!["course"],
      deleted_at = userMap.data()!["deleted_at"],
      email = userMap.data()!["email"],
      first_name = userMap.data()!["first_name"],
      gender = userMap.data()!["gender"],
      last_name = userMap.data()!["last_name"],
      phone = userMap.data()!["phone"],
      role_id = userMap.data()!["role_id"],
      school_id = userMap.data()!["school_id"],
      username = userMap.data()!["username"],
      year = userMap.data()!["year"],
      photoURL = userMap.data()!["photoURL"];
}

File: leap__backend/app/models/user.model.js
const sql = require("./db.js");

// constructor
const User = function(user) {
    this.uid = user.uid;
    this.first_name = user.first_name;
    this.last_name = user.last_name;
    this.username = user.username;
    this.email = user.email;
    this.phone = user.phone;
    this.gender = user.gender;
    this.address = user.address;
    this.birthday = user.birthday;
    this.school_id = user.school_id;
    this.role_id = user.role_id;
    this.course = user.course;
    this.year = user.year;
    this.photoURL = user.photoURL;
    this.deleted_at = user.deleted_at;
};

User.create = (newUser, result) => {
  sql.query("INSERT INTO users SET ?", newUser, (err, res) => {
    if (err) {
      console.log("error: ", err);
      result(err, null);
      return;
    }

    console.log("created user: ", { id: res.insertId, ...newUser });
    result(null, { id: res.insertId, ...newUser });
    return;
  });
};

User.findById = (uid, result) => {
  sql.query(`SELECT * FROM users WHERE uid = '${uid}'`, (err, res) => {
    if (err) {
      console.log("error: ", err);
      result(err, null);
      return;
    }

    if (res.length) {
      console.log("found users: ", res[0]);
      result(null, res[0]);
      return;
    }

    // not found Tutorial with the id
    result({ kind: "not_found" }, null);
    return;
  });
};

User.getAll = (result) => {
  let query = "SELECT * FROM users";

  sql.query(query, (err, res) => {
    if (err) {
      console.log("error: ", err);
      result(null, err);
      return;
    }

    console.log("users: ", res);
    result(null, res);
  });
};

User.updateById = (id, user, result) => {
  sql.query(
    `UPDATE users SET 
        first_name = ?, last_name = ?, username = ?, email = ?, phone = ?, gender = ?, address = ?, birthday = ?, school_id = ?, role_id = ?, course = ?, year = ?, photoURL = ?
    WHERE id = ?`,
    [
        user.first_name
        ,user.last_name
        ,user.username
        ,user.email
        ,user.phone
        ,user.gender
        ,user.address
        ,user.birthday
        ,user.school_id
        ,user.role_id
        ,user.course
        ,user.year
        ,user.photoURL
        , id
    ],
    (err, res) => {
      if (err) {
        console.log("error: ", err);
        result(null, err);
        return;
      }

      if (res.affectedRows == 0) {
        // not found Tutorial with the id
        result({ kind: "not_found" }, null);
        return;
      }

      console.log("updated users: ", { id: id, ...tutorial });
      result(null, { id: id, ...tutorial });
    }
  );
};

User.remove = (id, result) => {
  sql.query("DELETE FROM users WHERE id = ?", id, (err, res) => {
    if (err) {
      console.log("error: ", err);
      result(null, err);
      return;
    }

    if (res.affectedRows == 0) {
      // not found Tutorial with the id
      result({ kind: "not_found" }, null);
      return;
    }

    console.log("deleted users with id: ", id);
    result(null, res);
  });
};

module.exports = User;
